### ISSUE TITLE
Modifier.zoomable automatic snapback

### ISSUE THREAD

**Issue by vanniktech** [saket/telephoto#119]
Is there any way to automatically snap the zoom back when the user has stopped zooming? Meaning they have released their fingers from the screen.

So from this video as long as you can see that simulator helper tool for zooming in the zooming should work but when released, it should snap back to the original unzoomed state.

https://github.com/user-attachments/assets/d4acecfd-b33b-4bf3-b62a-817dc96f1e32

**Comment by saket** [comment]
Is [ZoomSpec#preventOverOrUnderZoom](https://github.com/saket/telephoto/blob/bf14bd6c0722da45b08a1d8c967ff71f1a5bd5f9/zoomable/src/commonMain/kotlin/me/saket/telephoto/zoomable/ZoomSpec.kt#L17-L23) what you're looking for?

```kotlin
val state = rememberZoomableState(
  zoomSpec = ZoomSpec(
    maxZoomFactor = 1f,
    preventOverOrUnderZoom = false,
  ),
)
```

**Comment by vanniktech** [comment]
This is perfect, yes. This is some special behavior though, right? Because in the documentation it says `maxZoomFactor` of 1f is no zooming at all, so it's a little bit confusing.

<img width="426" alt="Screenshot 2024-12-21 at 20 32 20" src="https://github.com/user-attachments/assets/322a0356-6ea6-4f08-a0f7-0fa123cce08a" />

**Comment by saket** [comment]
Ah, I can see how that might be confusing. I need to clarify that maxZoomFactor does not impact over-zooms. I'll try rewording this.

**Comment by vanniktech** [comment]
Let's say I have 3 images with a 1:1 aspect ratios underneath each other. Is there any trick to make each able to overlay each other? I currently have `Modifier.zIndex(Float.MAX_VALUE).zoomable(...)` which works from bottom to top (image can zoom over image above; but not below). I want both directions. Basically any image that is currently zoomed can overdraw all the others.

**Comment by saket** [comment]
Can you use `ZoomableState#zoomFraction` for determining their z-indices?

```kotlin
val isZoomedIn = state.zoomFraction > 0f // todo: handle null
Modifier.zIndex(if (isZoomedIn) 1f else 0f)
```

**Comment by vanniktech** [comment]
Nope, this does not work:

```kotlin
@Composable fun Modifier.magicZoom(): Modifier {
  val state = rememberZoomableState(
    zoomSpec = ZoomSpec(
      maxZoomFactor = 1f,
      preventOverOrUnderZoom = false,
    ),
  )
  val isZoomedIn = (state.zoomFraction ?: 0f) > 0f
  return zIndex(if (isZoomedIn) Float.MAX_VALUE else 1f).zoomable(
    state,
    clipToBounds = false,
  )
}
```

**Comment by saket** [comment]
Can you share more context about why that doesn't work? A video would also be useful.

**Comment by vanniktech** [comment]
Sure!

https://github.com/user-attachments/assets/744fbe12-73e8-4476-aae0-43c08a122a33

You can also try it here: https://github.com/vanniktech/playground/pull/303

**Comment by saket** [comment]
Oh I see, you're using a lazy layout, yea that won't work out of the box. IIRC `LazyLayout()` clips its items and doesn't offer any way to control their drawing order.

This will require a more sophisticated solution. If you want your zoomed image to draw over everything else, you'll possibly need to move your composable to an overlay backed by a separate window when a zoom gesture is detected. cc @ansman, who I remember was building something similar at his company earlier this year, and might have some tips to share.

**Comment by ansman** [comment]
We suffered from the same issue. To make matters worse even outside of a lazy column it still gets clipped by stuff like the app bar (which we didn't want). So instead we created a new `ComposeView` and added it to the decor view. You have to make sure the content is cached or that you notify the system when you are ready before swapping them to avoid flashes.

**Comment by vanniktech** [comment]
That sounds super complicated. I'll see if I can move around my functionality to avoid all of this.

**Comment by saket** [comment]
@ansman I've been passively wanting to offer a solution for this in `telephoto`. Do you have any code you could share from your overlay setup?

**Comment by saket** [comment]
@ansman and I worked on something together: [Modifier.zoomablePeekOverlay()](https://github.com/saket/telephoto/blob/99b24b99e93cd753a6245c5535f5c475cafffe91/zoomable-peek-overlay/src/main/kotlin/me/saket/telephoto/zoomable/ZoomablePeekOverlay.kt#L30) ([demo video](https://bsky.app/profile/saket.me/post/3leibuo3hqk2g)).

@vanniktech could you try it out using `0.15.0-SNAPSHOT` and share your feedback?

**Comment by vanniktech** [comment]
Is this typo here desired? https://github.com/saket/telephoto/blob/99b24b99e93cd753a6245c5535f5c475cafffe91/zoomable-peek-overlay/gradle.properties#L1C17-L1C39 

I've tried to use it: https://github.com/vanniktech/playground/pull/303/commits/4ace41bc07cc69cf3904d3316f3f536b803bdafc - but I'm not finding that new function.

**Comment by saket** [comment]
Oh whoops, can you try again?

```kotlon
implementation("me.saket.telephoto:zoomable-peek-overlay:0.15.0-SNAPSHOT")
```

**Comment by vanniktech** [comment]
This is Android only though, right? Which explains why I didn't find it. I've pushed another commit and now I'm getting:

```
Caused by: java.lang.ClassNotFoundException: Didn't find class "me.saket.telephoto.zoomable.ZoomLimit" on path: DexPathList[[zip file "/data/app/~~l9amg7Zah1rnP3PwdF-1dw==/app.playground.android.dev-gTIskoZUOYG7UUGI2EfYNg==/base.apk"],nativeLibraryDirectories=[/data/app/~~l9amg7Zah1rnP3PwdF-1dw==/app.playground.android.dev-gTIskoZUOYG7UUGI2EfYNg==/lib/arm64, /data/app/~~l9amg7Zah1rnP3PwdF-1dw==/app.playground.android.dev-gTIskoZUOYG7UUGI2EfYNg==/base.apk!/lib/arm64-v8a, /system/lib64, /system_ext/lib64]]
                                                                                                    	at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:259)
                                                                                                    	at java.lang.ClassLoader.loadClass(ClassLoader.java:637)
                                                                                                    	at java.lang.ClassLoader.loadClass(ClassLoader.java:573)
                                                                                                    	... 190 more
```

**Comment by saket** [comment]
Huh, I'm not sure why that is happening. Does it go away if you explicitly add the `zoomable` artifact?

```kotlin
implementation("me.saket.telephoto:zoomable:0.15.0-SNAPSHOT")
```

> This is Android only though, right?

Yes! I would love to add support for iOS, but I'll need help in translating this function for UIKit:

https://github.com/saket/telephoto/blob/8a14941074fe8f02e624042834f4ba220f188205/zoomable-peek-overlay/src/main/kotlin/me/saket/telephoto/zoomable/RealZoomablePeekOverlayState.kt#L78

**Comment by vanniktech** [comment]
I think something with the publishing went sideways. I had in `commonMain`: `implementation("me.saket.telephoto:zoomable:0.15.0-SNAPSHOT")` and then in `androidMain`: `me.saket.telephoto:zoomable-peek-overlay:0.15.0-SNAPSHOT`. Somehow this caused the exception to be thrown. When I removed the `zoomable` dependency on commonMain this was fixed.

Oh and yes, this is exactly what I want. It looks and feels really lit.
